gpflow/optimizers/mcmc.py

```diff
@@ -0,0 +1,118 @@
+# Copyright 2019 Artem Artemev @awav
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from dataclasses import dataclass
+from typing import Callable, List, Optional, TypeVar
+
+import tensorflow as tf
+
+from gpflow.base import Parameter, positive
+
+__all__ = ["positive_parameter", "SamplingHelper"]
+
+
+def positive_parameter(value: tf.Tensor):
+    if isinstance(value, (tf.Variable, Parameter)):
+        return value
+    return Parameter(value, transform=positive())
+
+
+ModelParameters = List[TypeVar("ModelParameter", tf.Variable, Parameter)]
+
+
+@dataclass(frozen=True)
+class SamplingHelper:
+    """
+    Helper reads from variables being set with a prior and writes values back to the same variables.
+
+    Example:
+        model = <Create GPflow model>
+        hmc_helper = SamplingHelper(lambda: -model.neg_log_marginal_likelihood(), model.trainable_parameters)
```

Comment:
`- neg_log_marginal_likelihood` seems a bit bizarre, I guess this is so that you properly include the log-prior (including log-det-jacobian of any transform)? would it perhaps be cleaner to refactor that in the Model base class ?